src/Subscriber.js

```diff
@@ -0,0 +1,382 @@
+
+        this.onUnsubscribeResponse = this.onUnsubscribeResponse.bind(this)
+        this.client.connection.on(ControlMessage.TYPES.UnsubscribeResponse, this.onUnsubscribeResponse)
+
+        this.onClientConnected = this.onClientConnected.bind(this)
+        this.client.on('connected', this.onClientConnected)
+
+        this.onClientDisconnected = this.onClientDisconnected.bind(this)
+        this.client.on('disconnected', this.onClientDisconnected)
+    }
+
+    onBroadcastMessage(msg) {
+        // Broadcast messages to all subs listening on stream-partition
+        const stream = this._getSubscribedStreamPartition(msg.streamMessage.getStreamId(), msg.streamMessage.getStreamPartition())
+        if (!stream) {
+            this.debug('WARN: message received for stream with no subscriptions: %s', msg.streamMessage.getStreamId())
+            return
+        }
+
+        const verifyFn = once(() => stream.verifyStreamMessage(msg.streamMessage)) // ensure verification occurs only once
+        // sub.handleBroadcastMessage never rejects: on any error it emits an 'error' event on the Subscription
+        stream.getSubscriptions().forEach((sub) => sub.handleBroadcastMessage(msg.streamMessage, verifyFn))
+    }
+
+    onSubscribeResponse(response) {
+        const stream = this._getSubscribedStreamPartition(response.streamId, response.streamPartition)
+        if (stream) {
+            stream.setSubscribing(false)
+            stream.getSubscriptions().filter((sub) => !sub.resending)
+                .forEach((sub) => sub.setState(Subscription.State.subscribed))
+        }
+        this.debug('Client subscribed: streamId: %s, streamPartition: %s', response.streamId, response.streamPartition)
+    }
+
+    onUnsubscribeResponse(response) {
+        this.debug('Client unsubscribed: streamId: %s, streamPartition: %s', response.streamId, response.streamPartition)
+        const stream = this._getSubscribedStreamPartition(response.streamId, response.streamPartition)
+        if (stream) {
+            stream.getSubscriptions().forEach((sub) => {
+                this._removeSubscription(sub)
+                sub.setState(Subscription.State.unsubscribed)
+            })
+        }
+
+        this._checkAutoDisconnect()
+    }
+
+    async onClientConnected() {
+        try {
+            if (!this.client.isConnected()) { return }
+            // Check pending subscriptions
+            Object.keys(this.subscribedStreamPartitions).forEach((key) => {
+                this.subscribedStreamPartitions[key].getSubscriptions().forEach((sub) => {
+                    if (sub.getState() !== Subscription.State.subscribed) {
+                        this._resendAndSubscribe(sub).catch((err) => {
+                            this.client.emit('error', err)
+                        })
+                    }
+                })
+            })
+        } catch (err) {
+            this.client.emit('error', err)
```

Comment:
Since `_resendAndSubscribe` has a catch above, seems like anything that would fall here would be a "genuine" error worth throwing?

Reply:
yep correct, good spotting